packages/truffle-decode-utils/src/types/types.ts

```diff
@@ -0,0 +1,541 @@
+import debugModule from "debug";
+const debug = debugModule("decode-utils:types");
+
+//type objects for Solidity types
+//these will just be defined as interfaces; there's not any particular need for
+//classes here (at least for now)
+
+//Note: This is NOT intended to represent every possible type representable by
+//a Solidity type identifier!  Only possible types of variables, not of all
+//possible values.  (Though there may be some expansion in the future.)
+//We do however count the builtin variables msg, block, and tx as variables
+//(not other builtins though for now) so there is some support for the magic
+//type.
+
+//We do include fixed and ufixed, even though these aren't implemented yet;
+//note though that values.ts does not include these.
+
+//NOTE: not all of these optional fields are actually implemented. Some are
+//just intended for the future.
+
+import BN from "bn.js";
+import { AstDefinition, AstReferences } from "../ast";
+import { Definition as DefinitionUtils } from "../definition";
+import { Contexts } from "../contexts";
+
+export namespace Types {
+
+  export type Type = UintType | IntType | BoolType | BytesType | AddressType
+    | FixedType | UfixedType | StringType | ArrayType | MappingType | FunctionType
+    | StructType | EnumType | ContractType | MagicType;
+
+  export interface UintType {
+    typeClass: "uint";
+    bits: number;
+  }
+
+  export interface IntType {
+    typeClass: "int";
+    bits: number;
+  }
+
+  export interface BoolType {
+    typeClass: "bool";
+  }
+
+  export type BytesType = BytesTypeStatic | BytesTypeDynamic;
+
+  export interface BytesTypeStatic {
+    typeClass: "bytes";
+    kind: "static";
+    length: number;
+  }
+
+  export interface BytesTypeDynamic {
+    typeClass: "bytes";
+    kind: "dynamic";
+    location?: string;
```

Comment:
So, they really should be, yes.  This also applies to certain other fields, in particular `visibility` and `mutability`.  The reason that they're not is because we often read them out of type identifiers.  But, I guess they could be enums if I just used a type coercion on reading them.  That would definitely make things nicer.  So, I'll try that, and unless something goes wrong I'll do it that way.

Reply:
OK, got it to build with enums, and surprisingly few type coercions :)